README.md
<p align="center"><img src="assets/oauth.png" width="200px" /></p>

# myanimelist-oauth

This is a simple Oauth 2.0 library for myanimelist API hosted on npm. This library can be used in browsers or node.

Note: all the bundles are in the dist directory, so you can pick whichever you want.

## Installation

```bash
yarn add myanimelist-oauth
```

## Initialization

```javascript
const malOauth = require("myanimelist-oauth");
```

## Usage

```javascript
const app = require("express")();
const malOauth = require("myanimelist-oauth");

app.get("/", (req, res) => {
	res.send("hello world");
});

var myanimelist = new malOauth({
	clientID: "your client ID",
	callback: "http://localhost:3000/callback",
	clientSecret: "your client secret", // not required if your application is registered as "Android", "IOS", or "Other"
});

app.get("/auth", myanimelist.authorize);

app.get("/callback", myanimelist.access_token, (req, res) => {
	res.json(req.get_data);
});

app.listen("3000", () => console.log("listening on port 3000"));
```

## Functions

```javascript
// Initializer
malOauth({
  clientID!: string,
  callback!: string,
  clientSecret?: string,
  codeChallange?: string,
  state?: string
})

authorize() => null // redirects to the uri you specified in your application registration

access_token() => null // middleware that sends final json response to req.get_data

refresh() => {
  'token-type': 'Bearer',
  'expires_in': int,
  'access_token': string,
  'refresh_token'string
}
```

## Contact

- Author: prince-ao
- Contact: princeaddo608@gmail.com


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [1.0.0] - 2021-12-17

### Added

- MalOauth class
- authorize function
- access_token function
- refresh function

## [1.0.2] - 2021-12-17

- fixed async bug in refresh function
- resized read me image


test/example.js
const app = require("express")();
const malOauth = require("../lib");
require("dotenv").config();

app.get("/", (req, res) => {
	res.send("hello world");
});

var myanimelist = new malOauth({
	clientID: process.env.CLIENT_ID,
	callback: process.env.CALLBACK,
	clientSecret: process.env.CLIENT_SECRET,
});

app.get("/auth", myanimelist.authorize);

app.get("/callback", myanimelist.access_token, (req, res) => {
	res.json(req.get_data);
});

app.listen("3000", () => console.log("listening on port 3000"));


lib/malOauth.js
const fetch = require("cross-fetch");

const _clientID = Symbol();
const _callback = Symbol();
const _codeChallenge = Symbol();
const _state = Symbol();
const _clientSecret = Symbol();
const _refreshToken = Symbol();

module.exports = class MalOauth {
	constructor(params) {
		if (params.clientID == "" || params.callback == "" || !params.clientID || !params.callback) {
			throw Error("clientID and callback are required");
		}
		const generateChallenge = () => {
			var result = "";
			var characters = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~";
			var charactersLength = characters.length;
			for (var i = 0; i < 128; i++) {
				result += characters.charAt(Math.floor(Math.random() * charactersLength));
			}
			return result;
		};
		this[_clientID] = params.clientID;
		this[_callback] = params.callback;
		this[_clientSecret] = params.clientSecret;
		this[_codeChallenge] = params.codeChallenge || generateChallenge();
		this[_state] = params.state;
	}
	authorize = (req, res) => {
		res.redirect(
			`https://myanimelist.net/v1/oauth2/authorize?response_type=code&client_id=${this[_clientID]}&code_challenge=${this[_codeChallenge]}&redirect_uri=${this[_callback]}&code_challenge_method=plain`
		);
		return;
	};

	access_token = async (req, res, next) => {
		const response = req.query;
		const rn = await fetch(`https://myanimelist.net/v1/oauth2/token`, {
			method: "POST",
			headers: {
				"Content-Type": "application/x-www-form-urlencoded",
			},
			body: `grant_type=authorization_code&code=${response.code}&client_id=${this[_clientID]}${
				this[_clientSecret] ? `&client_secret=${this[_clientSecret]}` : ""
			}&code_verifier=${this[_codeChallenge]}&redirect_uri=${this[_callback]}`,
		});

		const data = await rn.json();
		req.get_data = data;
		this[_refreshToken] = data.refresh_token;
		next();
	};

	refresh = async () => {
		if (!this[_refreshToken]) {
			throw Error("Could not find the refresh token. Try authorizing again.");
		}
		const rn = await fetch(`https://myanimelist.net/v1/oauth2/token`, {
			method: "POST",
			headers: {
				"Content-Type": "application/x-www-form-urlencoded",
			},
			body: `grant_type=refresh_token&client_id=${this[_clientID]}&client_secret=${this[_clientSecret]}&refresh_token=${this[_refreshToken]}`,
		});
		const data = await rn.json();
		return data;
	};
};


lib/oauth.js
var app = require("express")();
const fetch = require("cross-fetch");
//const fetch = require("node-fetch");

const _clientID = Symbol();
const _callback = Symbol();
const _codeChallenge = Symbol();
const _state = Symbol();
const _clientSecret = Symbol();

//app.use(bodyParser.json());

module.exports = class Oauth {
	constructor(params) {
		if (params.clientID == "" || params.callback == "" || !params.clientID || !params.callback) {
			throw Error("clientID, finishRedirect, and callback are required");
		}
		const generateChallenge = () => {
			var result = "";
			var characters = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~";
			var charactersLength = characters.length;
			for (var i = 0; i < 128; i++) {
				result += characters.charAt(Math.floor(Math.random() * charactersLength));
			}
			return result;
		};
		this[_clientID] = params.clientID;
		this[_callback] = params.callback;
		this[_clientSecret] = params.clientSecret;
		this[_codeChallenge] = params.code_challenge || generateChallenge();
		this[_state] = params.state;
	}
	authorize = (req, res) => {
		res.redirect(
			`https://myanimelist.net/v1/oauth2/authorize?response_type=code&client_id=${this[_clientID]}&code_challenge=${this[_codeChallenge]}&redirect_uri=${this[_callback]}&code_challenge_method=plain`
		);
		//console.log("code challenge in authorize is : " + this[_codeChallenge]);
		return;
	};

	access_token = async (req, res, next) => {
		const response = req.query;
		// console.log({
		// 	client_id: this[_clientID],
		// 	code: response.code,
		// 	code_challenge: this[_codeChallenge],
		// 	client_secret: this[_clientSecret],
		// 	grant_type: "authorization_code",
		// });
		//const body = `client_id=${this[_clientID]}&client_secret=${this[_clientSecret]}&code=${response.code}&code_verifier=${this[_codeChallenge]}&grant_type=authorization_code`;
		//console.log(body);
		//console.log("code challenge in access_token is : " + this[_codeChallenge]);
		res.redirect(`http://localhost:3000/test?code=${response.code}`);
		// fetch(`https://myanimelist.net/v1/oauth2/token`, {
		// 	method: "POST",
		// 	headers: {
		// 		"Content-Type": "application/x-www-form-urlencoded",
		// 	},
		// 	body: `grant_type=authorization_code&code=${response.code}&client_id=${this[_clientID]}&client_secret=${this[_clientSecret]}&code_verifier=${this[_codeChallenge]}`,
		// })
		// 	.then((res) => {
		// 		console.log(res);
		// 		res.json();
		// 	})
		// 	.then((json) => console.log(json));

		//const data = await rn.json();
		//console.log(rn);
		//res.json(rn);
	};

	get_data = async (req, res, next) => {
		const response = req.query.code;
		console.log({
			client_id: this[_clientID],
			code: response,
			code_challenge: this[_codeChallenge],
			client_secret: this[_clientSecret],
			grant_type: "authorization_code",
		});
		const res = await fetch(`https://myanimelist.net/v1/oauth2/token`, {
			method: "POST",
			headers: {
				"Content-Type": "application/x-www-form-urlencoded",
			},
			body: `grant_type=authorization_code&code=${response.code}&client_id=${this[_clientID]}&client_secret=${this[_clientSecret]}&code_verifier=${this[_codeChallenge]}`,
		});
	};
};
